tests/codegen-llvm/vec-into-iter-drops.rs

```diff
@@ -0,0 +1,28 @@
+//@ compile-flags: -Copt-level=3
+// Test that we can avoid generating an element dropping loop when `vec::IntoIter` is consumed.
+#![crate_type = "lib"]
+
+use std::vec;
+
+struct Bomb;
+impl Drop for Bomb {
+    #[inline]
+    fn drop(&mut self) {
+        panic!("dropped")
+    }
+}
+
+/// This test case from https://users.rust-lang.org/t/unnecessary-drop-in-place-emitted-for-a-fully-consumed-intoiter/135119
+/// It should not emit any `drop::<Bomb>()` because every element is forgotten.
+// CHECK-LABEL: @vec_into_iter_drop_option
+#[no_mangle]
+pub fn vec_into_iter_drop_option(v: vec::Vec<(usize, Option<Bomb>)>) -> usize {
+    // CHECK-NOT: panic
+    // CHECK-NOT: Bomb$u20$as$u20$core..ops..drop..Drop
```

Comment:
suggestion: `CHECK-NOT` tests are generally quite fragile, being prone to accidentally checking for something that never existed at all, or having the implementation in the standard library change such that a previously-useful test no longer does anything useful.

Thus I would suggest that whatever you're checking for here you also write another function that intentionally *does* trigger whatever you expect to not be there in this one, with *positive* `CHECK`s for the same things that are `CHECK-NOT`s here.

Reply:
Good point. However, I encounter a problem: I add more tests to this file and LLVM decides that they should call `drop_in_place::<IntoIter>()` instead of inlining it, defeating the entire point of this change. That’s a sign that this optimization is fragile and more work is needed, I guess. (Or maybe the perf results will show that it’s good despite that.)

I am now trying to write more explicit code that doesn’t rely on inlining to help.

Reply:
I’ve pushed a new commit 5fa286ddc6f88b9f309ea9e229a0e49071fd282b that performs explicit forgetting instead of relying on inlining and dead code elimination to achieve anything.